Free disk space in WebUI status bar

Piccirello opened the pull request:
> Puts free space on disk (Default save path) in WebUI's status 

This is a continuation of #6829 and addresses the need to poll for free disk space on a separate thread. Comments/suggestions on the QThread implementation are welcome.

<img width="376" alt="screen shot 2018-09-22 at 8 21 36 pm" src="https://user-images.githubusercontent.com/8296030/45922858-2ab22780-bea5-11e8-947b-94273eb7161f.png">

Closes #9680

thalieht commented on src/webui/freespacethread.h:
```diff
@@ -0,0 +1,52 @@
+
+private:
+    qint64 freeDiskSpace;
+    QTimer *timer;
```
"m_" prefixes.

Piccirello replied:
Will add these.

Piccirello commented:
I'm not a huge fan of passing `freeDiskSpace` as a parameter, but I haven't been able to think of another way for the static `btjson` methods to access the value.

Piccirello commented:
> btjson (as soon as prefjson) will cease to exist completely

That's right, I forgot about this change. Excellent!

Piccirello commented:
@Chocobo1 @glassez This has been rebased onto the latest web api changes.

Piccirello commented:
Please review when you get a chance :)

(will be addressing WebUI Search API comments this evening)

glassez commented:
>You quite misunderstood QThread. There is no thread in your code.
See our RSS::Parser for example of thread handling with Qt.

The main algorithm is:
1. Define worker class, e.g. FreeDiskSpaceChecker (subclass of QObject)
2. Create QThread object
3. Create worker object
4. Move worker object to your new thread (i.e. `moveToThread(yourThreadObject)`)
5. Connect worker signals/slots
6. Start your thread

The simply way is to define `FreeDiskSpaceChecker::update()` slot and `FreeDiskSpaceChecker::updated(qint64 freeSpaceSize)` signal. Then you can create some timer in SyncController and connect it to `FreeDiskSpaceChecker::update()` slot, and define some slot `freeDiskSpaceSizeUpdated(qint64 freeSpaceSize)` slot (where you just update corresponding SyncController variable) and connect it to `FreeDiskSpaceChecker::updated(qint64 freeSpaceSize)` signal.

Piccirello commented on src/app/application.cpp:
```diff
@@ -468,6 +469,7 @@ int Application::exec(const QStringList &params)
 
 #ifndef DISABLE_WEBUI
     m_webui = new WebUI;
+    FreeSpaceThread::initInstance();
```
It's only initialized during application startup.

glassez replied:
It's NOT "only initialized during application startup". It's implemented as Singleton, i.e. class with Globally accessed Single instance. I see no point to access it from any place except SyncController.

Piccirello commented on src/webui/api/synccontroller.cpp:
```diff
@@ -40,6 +40,7 @@
 #include "apierror.h"
 #include "isessionmanager.h"
 #include "serialize/serialize_torrent.h"
+#include "webui/freespacethread.h"
```
This is already an absolute path.

glassez commented on src/webui/api/synccontroller.cpp:
```diff
@@ -369,13 +371,26 @@ namespace
+        freeDiskSpaceChecker->moveToThread(m_freeDiskSpaceThread);
+        m_freeDiskSpaceThread->start();
+
+        m_freeDiskSpaceTimer = new QTimer(this);
```
Is there any strong reason to have timer here (not in FreeDiskSpaceChecker)?

glassez replied:
Looks like you can do without a timer. Just keep the time of the last update and compare it with the current one. When the difference exceeds the specified timeout, call update().

glassez commented on src/webui/api/synccontroller.cpp:
```diff
@@ -369,13 +371,26 @@ namespace
     auto lastResponse = sessionManager()->session()->getData(QLatin1String("syncMainDataLastResponse")).toMap();
     auto lastAcceptedResponse = sessionManager()->session()->getData(QLatin1String("syncMainDataLastAcceptedResponse")).toMap();
 
+    if (m_freeDiskSpaceThread == nullptr) {
```
nullptr_t is implicitly convertible to bool so we usually write it shorter: `if (m_freeDiskSpaceThread)`.

glassez commented on src/webui/api/synccontroller.h:
```diff
@@ -30,6 +30,9 @@
 
 #include "apicontroller.h"
 
+#include <QThread>
```
Forward declare both.

glassez commented on src/webui/api/synccontroller.h:
```diff
@@ -41,4 +44,10 @@ class SyncController : public APIController
 private slots:
     void maindataAction();
     void torrentPeersAction();
+	void freeDiskSpaceSizeUpdated(qint64 freeSpaceSize);
```
Indentation is broken here and below (all private section).

glassez commented on src/webui/freediskspacechecker.h:
```diff
@@ -0,0 +1,46 @@
+    Q_DISABLE_COPY(FreeDiskSpaceChecker)
+
+public:
+	FreeDiskSpaceChecker() {}
```
Fix indentation!
Use `= default;` instead of empty body.
Usually we keep `QObject *parent = nullptr` param for QObject subclasses.

Piccirello replied:
> Usually we keep QObject *parent = nullptr param for QObject subclasses.

Is there any point to this if the class isn't allowed to have a parent, due to being moved to a separate thread?

glassez commented on src/webui/freediskspacechecker.h:
```diff
@@ -0,0 +1,46 @@
+    void update();
+
+signals:
+    void updated(const qint64 freeSpaceSize);
```
Remove const from declaration.

glassez commented on src/webui/webui.pri:
```diff
@@ -11,6 +11,7 @@ HEADERS += \
     $$PWD/api/transfercontroller.h \
     $$PWD/api/serialize/serialize_torrent.h \
     $$PWD/extra_translations.h \
+    $$PWD/freediskspacechecker.h \
```
Why not in api folder? It's synccontroller helper.

Piccirello commented:
I previously tried `QTimer::singleShot()` and it ran in the same thread, but I'm guessing I must've messed something up. It's working as expected now.

Piccirello commented on src/webui/api/synccontroller.cpp:
```diff
@@ -369,13 +381,33 @@ namespace
     auto lastResponse = sessionManager()->session()->getData(QLatin1String("syncMainDataLastResponse")).toMap();
     auto lastAcceptedResponse = sessionManager()->session()->getData(QLatin1String("syncMainDataLastAcceptedResponse")).toMap();
 
+    if (!m_freeDiskSpaceThread) {
```
I don't love that this block of code exists in an api controller function. I'm almost considering switching back to a timer and moving all this to the `SyncController` constructor, so that this function only ever has to read the value in `m_freeDiskSpace`. @glassez thoughts?

Piccirello commented:
Minor refactor on `SyncController::maindataAction()` to remove the `if (!m_freeDiskSpaceThread)` `else` condition, so I think the code is a bit cleaner now. Rebased and ready for approval.

glassez commented on src/webui/extra_translations.h:
```diff
@@ -0,0 +1,98 @@
+/*
```
This file is removed from upstream. WebUI now has its own translations.

glassez commented on src/webui/api/synccontroller.h:
```diff
@@ -41,4 +45,13 @@ class SyncController : public APIController
+    void freeDiskSpaceSizeUpdated(qint64 freeSpaceSize);
+
+private:
+    ~SyncController();
```
Why is it private? You overthought it...

Piccirello commented on src/webui/api/freediskspacechecker.h:
```diff
@@ -0,0 +1,46 @@
+	FreeDiskSpaceChecker() = default;
+
+public slots:
+    void update();
```
Will update the names accordingly

Piccirello commented:
All comments addressed and squashed. I moved the init logic to the `SyncController` constructor and the `FreeDiskSpaceChecker::check` call to a separate `SyncController::getFreeDiskSpace()` function.

glassez commented on src/webui/api/freediskspacechecker.h:
```diff
@@ -0,0 +1,46 @@
+    Q_DISABLE_COPY(FreeDiskSpaceChecker)
+
+public:
+    FreeDiskSpaceChecker() = default;
```
You can have no such declaration. Since there are no other constructors defined compiler will generate the default one itself.

Piccirello replied:
Without the explicit constructor I get this error

```
webui/api/synccontroller.cpp: In constructor ‘SyncController::SyncController(ISessionManager*, QObject*)’:
webui/api/synccontroller.cpp:323:55: error: no matching function for call to ‘FreeDiskSpaceChecker::FreeDiskSpaceChecker()’
     m_freeDiskSpaceChecker = new FreeDiskSpaceChecker();
```

glassez replied:
Oops... Q_DISABLE_COPY really creates copy constructor (either private or deleted).

glassez replied:
Oops... Q_DISABLE_COPY really creates copy constructor (either private or deleted).

glassez replied:
Oops... Q_DISABLE_COPY really creates copy constructor (either private or deleted).

glassez replied:
Oops... Q_DISABLE_COPY really creates copy constructor (either private or deleted).

glassez replied:
Oops... Q_DISABLE_COPY really creates copy constructor (either private or deleted).

glassez commented on src/webui/api/synccontroller.cpp:
```diff
@@ -312,6 +316,26 @@ namespace
+
+SyncController::~SyncController()
+{
+    if (m_freeDiskSpaceThread) {
```
Now this check is redundant since you create it in constructor.

glassez commented on src/webui/api/synccontroller.cpp:
```diff
@@ -482,3 +508,18 @@ void SyncController::torrentPeersAction()
+qint64 SyncController::getFreeDiskSpace()
+{
+    const qint64 now = QDateTime::currentMSecsSinceEpoch();
+    if ((now - m_freeDiskSpaceLastUpdate) >= 30000) {
```
Please define named constant for update timeout.

glassez commented on src/webui/api/synccontroller.h:
```diff
@@ -38,7 +44,19 @@ class SyncController : public APIController
     using APIController::APIController;
 
+    explicit SyncController(ISessionManager *sessionManager, QObject *parent = nullptr);
+    ~SyncController();
```
Add `override`.

glassez commented on src/webui/api/synccontroller.h:
```diff
@@ -38,7 +44,19 @@ class SyncController : public APIController
+    qint64 getFreeDiskSpace();
+
+    qint64 m_freeDiskSpace = 0;
+    FreeDiskSpaceChecker *m_freeDiskSpaceChecker;
```
All other fields have default value. Why doesn't this one?

glassez commented:
@Chocobo1, please take a look at this PR, it's almost finished.

Chocobo1 commented on src/webui/api/synccontroller.cpp:
```diff
@@ -312,6 +316,26 @@ namespace
+    m_freeDiskSpaceThread = new QThread(this);
+    m_freeDiskSpaceChecker = new FreeDiskSpaceChecker();
+    m_freeDiskSpaceChecker->moveToThread(m_freeDiskSpaceThread);
+    m_freeDiskSpaceThread->start();
```
I would prefer moving this line under the 2 `connect()`.

Chocobo1 commented on src/webui/api/freediskspacechecker.cpp:
```diff
@@ -0,0 +1,38 @@
+/*
```
@glassez 
Should this file placed under src/base (or src/base/utils) or it is fine in src/webui/api?

glassez replied:
Let's keep it here since it isn't used outside of WebAPI. Or maybe move to `src/webui/api/private` to distinguish from api controllers. I don't think it will be needed somewhere else.

glassez commented on src/webui/api/synccontroller.h:
```diff
@@ -38,7 +44,19 @@ class SyncController : public APIController
+    qint64 m_freeDiskSpace = 0;
+    FreeDiskSpaceChecker *m_freeDiskSpaceChecker;
+    QThread *m_freeDiskSpaceThread = nullptr;
+    qint64 m_freeDiskSpaceLastUpdate = 0;
```
Why we need yet another timer? What real benefit do you suppose to get?

Chocobo1 replied:
ah yes, some over thinking, forget it.

Chocobo1 replied:
>Why we need yet another timer? What real benefit do you suppose to get?

Out of curiosity I looked up its source, it is not the usual QTimer and it is quite lightweight.

IMO the most benefit is a specific type to represent the task here: counting elapsed time.
It still would be nice to use it, but I don't insist.

glassez replied:
>Out of curiosity I looked up its source, it is not the usual QTimer and it is quite lightweight.

If it doesn't really use system timers I don't object to use it. Of course, if @Piccirello wants to change it one more time again.

glassez replied:
>Out of curiosity I looked up its source, it is not the usual QTimer and it is quite lightweight.

If it doesn't really use system timers I don't object to use it. Of course, if @Piccirello wants to change it one more time again.

Piccirello replied:
Switched to using `QElapsedTimer`. This type does make the intentions of the code very clear.

Piccirello commented:
I agree with all comments, all addressed and squashed (with exception of the default constructor issue).

glassez commented on src/webui/api/synccontroller.cpp:
```diff
@@ -482,3 +507,19 @@ void SyncController::torrentPeersAction()
+qint64 SyncController::getFreeDiskSpace()
+{
+    const qint64 now = QDateTime::currentMSecsSinceEpoch();
+    const auto thirtySeconds = 30000;
```
When I ask you to create named constant for it I meant something other and I hoped it's so obvious to understand it without any additional description.
Usually we create such "compile time configuration" constants at top of the file for easily changing it in the future without searching through all the code. And it should have an appropriate name to easily understand what is it configure (e.g. FREEDISKSPACE_CHECK_TIMEOUT).
It's absolutely pointless to have it at this place.

Piccirello replied:
> It's absolutely pointless to have it at this place.

Not totally pointless, it clearly documents the length of time between calculation. I will move it to the top of the file.

Piccirello replied:
Moved this constant to the top of `SyncController` and squashed

Piccirello commented on src/webui/api/synccontroller.h:
```diff
@@ -28,6 +28,8 @@
 
 #pragma once
 
+#include <QElapsedTimer>
```
When I try forward declaring the class (and moving the include to the .cpp) I get this error:

```
compiling moc_synccontroller.cpp
In file included from webui/webapplication.cpp:62:0:
webui/api/synccontroller.h:62:19: error: field ‘m_freeDiskSpaceElapsedTimer’ has incomplete type
     QElapsedTimer m_freeDiskSpaceElapsedTimer;
```

glassez replied:
You can't create variable (class field here) of incomplete class. Compiler need to know its definition to do it.

glassez commented:
Thank you!
